tools/generator/src/Extensions/XcodeScheme+Extensions.swift

```diff
@@ -0,0 +1,170 @@
+// MARK: TargetWithID
+
+extension XcodeScheme {
+    struct TargetWithID {
+        let id: TargetID
+        let target: Target
+    }
+}
+
+extension XcodeScheme.TargetWithID: Equatable {
+    static func == (lhs: XcodeScheme.TargetWithID, rhs: XcodeScheme.TargetWithID) -> Bool {
+        return lhs.id == rhs.id
+    }
+}
+
+extension XcodeScheme.TargetWithID: Comparable {
+    static func < (lhs: XcodeScheme.TargetWithID, rhs: XcodeScheme.TargetWithID) -> Bool {
+        return lhs.target.platform < rhs.target.platform
+    }
+}
+
+// MARK: Resolve TargetIDs
+
+extension XcodeScheme {
+    /// Represents a configuration string (Target.configuration).
+    typealias Configuration = String
+
+    /// Determines the mapping of `LabelValue` to the `TargetID` values based upon the scheme's
+    /// configuration.
+    func resolveTargetIDs(targets: [TargetID: Target]) throws -> [LabelValue: TargetID] {
+        // Get all of the scheme labels
+        let allSchemeLabels = allSchemeLabels
+        let topLevelSchemeLabels = allSchemeLabels.filter(\.isTopLevel)
+        let otherSchemeLabels = allSchemeLabels.subtracting(topLevelSchemeLabels)
+
+        // Gather target info for the top-level targets
+        let topLevelLabels = Set(topLevelSchemeLabels.map(\.label))
+        let topLevelTargetWithIDs = targets
+            .filter { _, target in topLevelLabels.contains(target.label) }
+            .map { id, target in TargetWithID(id: id, target: target) }
+        let topLevelTargetInfoByLabelValue = collectTargetInfoByLabelValue(
+            targetWithIDs: topLevelTargetWithIDs
+        )
+
+        var resolvedTargetIDs = [LabelValue: TargetID]()
+
+        // Collect top-level targetIDs
+        var topLevelTargetIDs = Set<TargetID>()
+        var topLevelPlatforms = Set<Platform>()
+        for schemeLabel in topLevelSchemeLabels {
+            guard let targetInfo = topLevelTargetInfoByLabelValue[schemeLabel.label] else {
+                throw PreconditionError(message: """
+Did not find `targetInfo` for top-level label "\(schemeLabel.label)"
+""")
+            }
+            let targetID = try targetInfo.best().id
+            topLevelTargetIDs.update(with: targetID)
+            topLevelPlatforms.formUnion(targetInfo.platforms)
+            resolvedTargetIDs[schemeLabel.label] = targetID
+        }
+
+        // Reduce the number of targets being evaluated. Only include those that have one of the
+        // top-level platforms.
+        let targetsWithTopLevelPlatforms = targets.filterDependencyTree(
+            startingWith: topLevelTargetIDs
+        ) { target in
+            topLevelPlatforms.contains(target.platform)
+        }
```

Comment:
@brentleyjones This filters the set of targets being evaluated to those that are in the build tree and has a platform that is one of the top-level platfoms.

Reply:
Yes. I thought that applying the filter upfront made the code a little easier to understand as it simplifies the label check below. Also, I was using the `multiplatform` example as a use case. Filtering the targets list once does reduce the number of iterations that we potentially need to do when we perform the label search below.